streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1938,115 +1938,134 @@ object ZChannel {
     mergeStrategy: => MergeStrategy = MergeStrategy.BackPressure
   )(
     f: (OutDone, OutDone) => OutDone
-  )(implicit trace: Trace): ZChannel[Env, InErr, InElem, InDone, OutErr, OutElem, OutDone] =
+  )(implicit trace: Trace): ZChannel[Env, InErr, InElem, InDone, OutErr, OutElem, OutDone] = {
+    sealed trait Result
+    object Result {
+      final case class Value(value: OutElem)        extends Result
+      final case class Error(error: OutErr)         extends Result
+      final case class Done(done: OutDone)          extends Result
+      final case class Fatal(cause: Cause[Nothing]) extends Result
+
+      def value(v: OutElem): Result            = Value(v)
+      def error(e: OutErr): Result             = Error(e)
+      def done(d: OutDone): Result             = Done(d)
+      def fatal(cause: Cause[Nothing]): Result = Fatal(cause)
```

Comment:
Do we need these methods? Can't we use the apply method of the case classes directly?

Reply:
No we don't :)